### ISSUE TITLE
Resolving system dependency conflicts (meta)

### ISSUE THREAD

**Issue by stonier** [RobotLocomotion/drake#7412]
**FCL & Dependencies**

Strategy is to hide symbols in FCL & CCD. Octomap will use the same version as that in ROS (1.8.1) as opposed to system. Most users are likely on the same version, if this changes, then we can rethink/hide symbols there. Note that Drake only uses octomap indirectly via FCL.

* [x] **ccd** : send patch upstream and merged for symbol hiding (~~[libccd/pull/39](https://github.com/danfis/libccd/pull/39)~~)
* [x] **ccd** : wait for patch to be accepted
* [x] **fcl** : send patch upstream and merged for symbol hiding (~~[fcl/pull/233](https://github.com/flexible-collision-library/fcl/pull/233)~~)
* [x] **fcl bugfixing** : reported a discovered bug upstream, but does not affect us directly (~~[fcl/issues/244](https://github.com/flexible-collision-library/fcl/issues/244)~~)
* [ ] **ccd/fcl/octomap** : drake update for the holy trinity ([drake/pull/7417](https://github.com/RobotLocomotion/drake/pull/7417))

**YAML**

* [x] Drop back to system yaml, (~~[drake/pull/7352](https://github.com/RobotLocomotion/drake/pull/7352)~~)
* [x] Await system protobuf integration (~~[drake/pull/7323](https://github.com/RobotLocomotion/drake/pull/7323)~~)
* [x] Await crosstool fixes for Mac (~~[drake/pull/7404]~~(https://github.com/RobotLocomotion/drake/pull/7404))
* [ ] Resubmit PR (see ~~[revert PR](https://github.com/RobotLocomotion/drake/pull/7403)~~, [drake/pull/7602](https://github.com/RobotLocomotion/drake/pull/7602))

**VTK**

* [x] Privatise and do not install vtk_utils.h, (~~[drake/pull/7258](https://github.com/RobotLocomotion/drake/pull/7258/files)~~)

**PCL Enabling**

* [x] **eigen** : workaround PCLConfig.cmake issues with Eigen, ([shambhala#38](https://github.com/RobotLocomotion/drake-shambhala/issues/38)) (~~[drake/pull/7435](https://github.com/RobotLocomotion/drake/pull/7435)~~)([shambhala/pull/65](https://github.com/RobotLocomotion/drake-shambhala/pull/65))
* [x] **vtk** : resolve a vtk linking bug that occurs when compiling against drake & pcl (~~[drake/pull/7434](https://github.com/RobotLocomotion/drake/pull/7434)~~)

**Protobuf**

* [x] initial idea and working notes (~~[drake/pull/7278](https://github.com/RobotLocomotion/drake/pull/7278)~~)
* [x] solve ~~[matlab 2017a](https://drake-jenkins.csail.mit.edu/job/linux-xenial-gcc-cmake-experimental-matlab/53/consoleFull)~~ problems
* [x] await homebrew python fixes, ~~#7440~~
* [x] fall back to system protobuf 2.6 (~~[drake/pull/7323](https://github.com/RobotLocomotion/drake/pull/7323)~~)
* [ ] decouple drake from protobuf (optional)

More detailed notes in this [spreadsheet](https://docs.google.com/spreadsheets/d/1tiSN8LtNCgos-Vt3U4IBt8okEIC5CKiYLdLx0yo9EfU/edit#gid=863450394).

@nuclearsandwich, @j-rivero, @clalancette can you add comments as things get resolved, or new items are brought to the fore. I'll update this list as we go.

**Comment by jwnimmer-tri** [comment]
FYI I reduced the permissions on the "spreadsheet" link above from "Anyone with the link can edit" to "Anyone with the link can view".  Otherwise, the document would turn into a w4r3z site.  We may have to add back individually-authenticated commentors or editors later to compensate.

**Comment by stonier** [comment]
Thanks

**Comment by stonier** [comment]
From the `WORKSPACE` file, it looks as if we are actually not currently using forks of `fcl` nor `libccd`, just hashes from the original projects. They both certainly are getting upstream patches.

**libccd**

We are in fact using a very old tag of libccd (2014). There has been very little code changes since then - just one commit I think. The rest have been in CMake upgrade and it looks like @jamiesnape has already started to do some symbol hiding. Diff between tip and v2.0 [here](https://github.com/danfis/libccd/compare/v2.0...master). Since it already has some logic for exports in tip, we should probably try and upgrade Drake to use that and do whatever patching is necessary on top of that to hide the symbols in Drake.

* @j-rivero : check that Drake builds from tip
* @j-rivero : if that works, create a PR that points to tip of the [RobotLocomotion/libccd](https://github.com/RobotLocomotion/libccd) fork I just created
* @avalenzu : to follow up and make sure that the contact collisions still work with the PR
* @j-rivero : if necessary, drop a patch on the fork for hiding symbols, update the PR, and upstream the patch

Sound like a plan?

**FCL**

It has a fork in [RobotLocomotion/fcl](https://github.com/RobotLocomotion/fcl) now.

**Comment by j-rivero** [comment]
> Sound like a plan?

Yep. Working on it.

**Comment by j-rivero** [comment]
@stonier the pull request should be ready at pull request #7417 

The fcl cherry pick for your own fork is at https://github.com/RobotLocomotion/fcl/pull/1

**Comment by stonier** [comment]
Some updates to the above.

**Yaml** 

Yaml is blocking on system protobuf getting in because of a Mac issue on crosstool not having `/usr/local/include` which indirectly itself causes issues with the protobuf 3.1 currently being used. Refer to the causal chain in the first post.

**Protobuf**

Pending on a solution to adopting homebrew's python. @jamiesnape can you raise that issue and start discussion on how to solve this ASAP?

**Comment by jamiesnape** [comment]
> Pending on a solution to adopting homebrew's python. @jamiesnape can you raise that issue and start discussion on how to solve this ASAP?

#7440

**Comment by jamiesnape** [comment]
> We are in fact using a very old tag of libccd (2014). There has been very little code changes since then - just one commit I think. The rest have been in CMake upgrade and it looks like @jamiesnape has already started to do some symbol hiding. Diff between tip and v2.0 here. Since it already has some logic for exports in tip, we should probably try and upgrade Drake to use that and do whatever patching is necessary on top of that to hide the symbols in Drake.

You should be able to use all the logic I added to hide everything that you need to and keep their three build systems working rather than trying to do so with CMake-specific logic per https://github.com/danfis/libccd/pull/39. Not using `GenerateExportHeader` was deliberate given the requirements of that project, though maybe that changed.

**Comment by nuclearsandwich** [comment]
> - [ ] eigen : workaround PCLConfig.cmake issues with Eigen, (shambhala#38) https://github.com/RobotLocomotion/drake/pull/7435
> - [x] vtk : resolve a vtk linking bug that occurs when compiling against drake & pcl https://github.com/RobotLocomotion/drake/pull/7434

**Comment by nuclearsandwich** [comment]
https://github.com/RobotLocomotion/drake/pull/7453 excludes the yaml_util header from the installation without reverting back to system yaml-cpp. We probably still want to re-introduce the system yaml-cpp changes once the build on mac stabilizes.

**Comment by j-rivero** [comment]
> Not using GenerateExportHeader was deliberate given the requirements of that project, though maybe that changed.

To pass all the CI tests (they are using cmake, autotools and Makefile) you are right, we can not rely only in the `GenerateExportHeaders` of cmake. I've changed the PR to use an static file for the exports and patch Makefile and autotools. All tests passes now.

**Comment by jamiesnape** [comment]
To cut down the size of the diff can you keep the macros named `_ccd_export` and keep them in `compiler.h`. Also you can keep the `project(libccd C)` line now, you have a reference to FCL instead of CCD in one of the comments, and their CMake style is to have a continued line indent of two and no space before opening parentheses in the files.

**Comment by jamiesnape** [comment]
Looks like libccd merged the patch.

**Comment by jwnimmer-tri** [comment]
I think this might be finished?  At least, the checkbox list in the top post shows that everything has merged.

**Comment by nuclearsandwich** [comment]
Every dependency with an identified conflict was resolved. So yes, unless/until new issues are reported I think this could be closed.

**Comment by jwnimmer-tri** [comment]
Thanks!  This was very exciting to see happen, it should be a nice improvement in compatibility.  I will close for now but we can re-open and/or file a new issue if we find we need other improvements.
